### ISSUE TITLE
array_splice to get an url's parameters is wrong

### ISSUE THREAD

**Issue by nkovacs** [yiisoft/yii2#7894]
If you have a url that is an array, and you add the route later, even though the route will be `$url[0]`, its offset won't be 0, so using `array_splice($url, 1)` to get the parameters is wrong.

For example:

```php
$url = ['param1' => 'foo', 'param2' => 'bar'];
$url[] = 'controller/action';

$params = array_splice($url, 1); // will be ['param2' => 'bar', 0 => 'controller/action']
```

* https://github.com/yiisoft/yii2/blob/bbbeae95a15f6299dae202f2c64367b49ee0400f/framework/web/Application.php#L78
* https://github.com/yiisoft/yii2/blob/81bcd116a83dc24c9db2d8cd73f65074e09863bf/framework/widgets/Menu.php#L318
* https://github.com/yiisoft/yii2-bootstrap/blob/84ad12bc5c409d62db1cef6e25023beccf3b8f1a/Nav.php#L244

**Comment by samdark** [comment]
Can you provide a way to reproduce it as a user?

**Comment by nkovacs** [comment]
I haven't run into this, so I can only come up with contrived examples. But that doesn't change the fact that it's wrong.

```php
// in controller
$url = array_intersect_key($_GET, array_flip($allowedParams));
$url[] = 'some-action';

// then use $url in menu or nav
```

or this, though it's unlikely someone would write the url like this:

```php
'catchAll' => ['reason' => 'maintenance', 'site/offline'],
```

**Comment by samdark** [comment]
Well, these are definitely artificial examples of doing things wrong i.e. not reading docs. I'm asking not about developer mistakes but about user playing with browser address bar.

**Comment by nkovacs** [comment]
I based the first example on Url::canonical:
https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseUrl.php#L304

**Comment by klimov-paul** [comment]
Resolved by commit 91a6e4b

**Comment by klimov-paul** [comment]
`yii\bootstrap\Nav` fixed by https://github.com/yiisoft/yii2-bootstrap/commit/b91b32530e5ba7e12b4d1b07aafd2f156d5599c7
